test/e2e/upgrades/apps/daemonsets.go

```diff
@@ -118,22 +122,49 @@ func (t *DaemonSetUpgradeTest) validateRunningDaemonSet(f *framework.Framework)
 	}
 }
 
-func checkRunningOnAllNodes(f *framework.Framework, namespace string, selector map[string]string) (bool, error) {
+// newPod returns a pod based on DaemonSet's configuration. This's a shadow helper function
+// of `pkg/controller/daemon/util.NewPod`.
+func newPod(ds *extensions.DaemonSet, nodeName string) *v1.Pod {
+	newPod := &v1.Pod{Spec: ds.Spec.Template.Spec, ObjectMeta: ds.Spec.Template.ObjectMeta}
+	newPod.Namespace = ds.Namespace
+	newPod.Spec.NodeName = nodeName
+
+	// Added default tolerations for DaemonSet pods.
+	util.AddOrUpdateDaemonPodTolerations(&newPod.Spec, kubelettypes.IsCriticalPod(newPod))
+
+	return newPod
+}
+
+func canScheduleOnNode(node v1.Node, ds *extensions.DaemonSet) bool {
+	newPod := newPod(ds, node.Name)
+	nodeInfo := schedulercache.NewNodeInfo()
+	nodeInfo.SetNode(&node)
+	fit, _, err := daemon.Predicates(newPod, nodeInfo)
+	if err != nil {
+		framework.Failf("Can't test DaemonSet predicates for node %s: %v", node.Name, err)
+		return false
+	}
+
+	return fit
+}
+
+func checkRunningOnAllNodes(f *framework.Framework, ds *extensions.DaemonSet) (bool, error) {
 	nodeList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
 	if err != nil {
 		return false, err
 	}
 
 	nodeNames := make([]string, 0)
 	for _, node := range nodeList.Items {
-		if len(node.Spec.Taints) == 0 {
-			nodeNames = append(nodeNames, node.Name)
-		} else {
-			framework.Logf("Node %v not expected to have DaemonSet pod, has taints %v", node.Name, node.Spec.Taints)
+		if !canScheduleOnNode(node, ds) {
+			framework.Logf("DaemonSet pods can't schedule to node %s, skip checking this node", node.Name)
+			continue
 		}
+
+		nodeNames = append(nodeNames, node.Name)
```

Comment:
b/c e2e's can be run in the wild w/varying policy shouldn't there be a check for >=1 node otherwise error?